### PR TITLE
WIP: build `openswoole` extension from sources, as part of `docker build`

### PR DESCRIPTION
**Pull request by Ocramius**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | yes
| QA            | yes

### Description

While working on #71, I found out that the docker layers responsible for installing `swoole` and `openswoole` amount for ~400Mb of wasted image size.

In practice, the size wasted is probably not avoidable, since `openswoole.so` is ~40Mb (!!!), and we need one `.so` file per-php-version, which is a lot of duplication.

While investigating that, I discovered the approach used by @weierophinney to just "paste a `.tar.gz` file into the image", which, while quite clever, is really a mess, and depends on @weierophinney's manual intervention, should we need new `openswoole` versions.

Hereby I created a new build stage dedicated to assembling an `openswoole.so` from source.

Note: worth merging #71 first, so this diff becomes more readable/understandable. 

TODOs:

 * [ ] build also `swoole.so`
 * [ ] use the `openswoole.so` resulting binaries in the final image
 * [ ] get rid of the `.tar.gz` download "hack"
 * [ ] perhaps make `openswoole.so` installation opt-in, since it is so extremely gigantic?

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,55 +1,266 @@
...
+RUN cd /markdownlint \
+    && npm ci \
+    # Smoke-testing the installation, just making sure it works as expected - should pass first file, fail on second
+    && node_modules/.bin/markdownlint-cli2 /test-files/dummy-ok-markdown-test-file.md \
+    && if node_modules/.bin/markdownlint-cli2 /test-files/dummy-ko-markdown-test-file.md; then exit 1; else exit 0; fi
 
-# Cleanup
-RUN rm -rf /setup \
+
+FROM base-distro AS install-openswoole
```
Comment (Ocramius) [inline Dockerfile:31]: This build stage is the actual relevant part of this patch

File: Dockerfile [hunk 2]
```diff
@@ -1,55 +1,266 @@
...
-RUN mkdir -p /usr/local/share/composer \
-    && composer global require staabm/annotate-pull-request-from-checkstyle \
-    && ln -s /usr/local/share/composer/vendor/bin/cs2pr /usr/local/bin/cs2pr
+
+# Build/install static modules that do not have packages
+COPY mods-available /mods-available
+COPY mods-install /mods-install
+RUN /mods-install/install_sqlsrv.sh
+RUN /mods-install/install_swoole.sh
+#RUN for INSTALLER in /mods-install/*.sh;do ${INSTALLER} ; done
```
Comment (Ocramius) [inline Dockerfile:263]: I split this step just to investigate the size of the various layers

**Comment by weierophinney** [comment]
My big question is: what does this do to the build times for *this* container?

That was the big reason I offloaded the build elsewhere: it was making the build take upwards of 30 minutes, which makes evaluating PRs ridiculous, particularly if you run into issues and need to push multiple commits to fix them.

From what I can see in the build details here... it didn't even attempt to build openswoole. We need to get it to do that to get an idea of how long it will take to see if this is feasible.

Alternately, I could move the download and install part into the mezzio-swoole library itself, and/or break up the `.so` and `.ini` files into discrete packages per PHP version, which would allow us to have an install rule for swoole itself that fetched them when they are requested.

**Comment by Ocramius** [comment]
> My big question is: what does this do to the build times for this container?

Depends: probably 30m without caching, but consider that the build stage is made independent from the rest 

> That was the big reason I offloaded the build elsewhere: it was making the build take upwards of 30 minutes, which makes evaluating PRs ridiculous, particularly if you run into issues and need to push multiple commits to fix them.

That's precisely what build stages do 😁

Alternatively, we can pull the `.so` files out of the docker images published by `openswoole`, or use their PPAs (only PHP 7.4+), but there is nothing for `swoole.so`

**Comment by internalsystemerror** [comment]
I recommend using https://github.com/mlocati/docker-php-extension-installer in order to install the extensions. It already covers `swoole` and `openswoole` for the PHP versions used here.

Example usage:
```Dockerfile
COPY --from=mlocati/php-extension-installer /usr/bin/install-php-extensions /usr/local/bin/

RUN install-php-extensions gd swoole openswoole intl
```

This handles the OS and PHP version requirements and generally makes the builds a lot simpler.

**Comment by Ocramius** [comment]
That would certainly be interesting to use in a pre-run step, defined by downstream repositories.

If we do that, then we can reduce this docker image by ~400Mb, perhaps more, considering that we probably won't need Microsoft's ODBC libraries and `phpX-dev` dependencies pre-installed. 

Still, that would require a new major version of this container.

**Comment by weierophinney** [comment]
> Still, that would require a new major version of this container.

Since installation/enablement of extensions is handled via the container, and this is done in a script... we absolutely _could_ require the mlocati installer in the container, and have the script call it to install those extensions when the repo is configured to use them. That would prevent the need for a new minor, because the container would still work with the same configuration.

**Comment by boesing** [comment]
Can I be of any help here?

**Comment by Ocramius** [comment]
If we still want openswoole as part of the container: do we actually need it? It raises container size by ~40Mb * number of supported PHP versions just to exist AFAIK

**Comment by boesing** [comment]
Hm, I liked the idea of that `mlocati/docker-php-extension-installer` but it requires the official php docker image https://github.com/mlocati/docker-php-extension-installer/blob/9a11492f59b9fac306c1009445e62484d97a1eb2/install-php-extensions#L16

It might be somehow possible to make it work by importing the mandatory files from https://github.com/docker-library/php and modifying them to our needs but I don't like the idea to actually apply patches to a 3rd-party script.

**Comment by Ocramius** [comment]
Another alternative could be to run a specific container image, depending on PHP version/matrix selected.

For example `nixery.dev` does that with a fancy `docker run -ti nixery.dev/shell/git/htop bash` (which means installing `shell`, `git`, `htop`).

Anyway, wishful thinking :D 

If a larger docker image is not **yet** a problem, we can roll with this current approach.

**Comment by weierophinney** [comment]
Another possibility: I could switch the logic to pull the pre-built extension when it has been requested via a workflow. That would reduce the container size and speed up builds.

Ideally, if we went that route, I'd also have a package per PHP version, since it would be requested for a single pho version at a time, and doing so would speed up downloads when it's been requested.

Thoughts?

**Comment by Ocramius** [comment]
@weierophinney I didn't fully understand this: 

 > Ideally, if we went that route, I'd also have a package per PHP version, since it would be requested for a single pho version at a time, and doing so would speed up downloads when it's been requested.

Are you proposing to publish swoole pre-built packages in here (as artifacts)?

**Comment by weierophinney** [comment]
No - I'd do it in the repo where I'm currently building the Swoole/OpenSwoole extension bundle. We'd continue pulling from there, but:

- we'd download when a workflow requests to enable that extension, and
- we'd download the extension built for that specific php version

**Comment by boesing** [comment]
Can we probably outline again on why we need to prebuild `swoole`?
As far as I could see, sury provides `swoole` extension already and thus - it can be installed the same way as all the other extensions we do try to install when they're missing (i.e. via `apt install phpX.Y-swoole`)

So we only need `openswoole` here, or are there specific quirks provided via your repository @weierophinney?


> we'd download when a workflow requests to enable that extension

Imho my preferred way because thats exactly what we do with all the other (missing) extensions.
I already implemented `composer.json` parsing for PHP  extensions as well.

**Comment by weierophinney** [comment]
At the time we started testing against Swoole, the sury repo did NOT have packages, which is what prompted the original solution. He's only added those since PHP 8.1, from what I can tell, and I need to verify that he has them for all PHP versions we're currently supporting (that's the first concern). Then, as you noted, openswoole is the other issue, as he is not providing that extension at all (which is interesting, as most of the active development in that space has moved to openswoole).

So, action items:

- Determine if the Sury repo's swoole packages exist for all PHP versions we've been providing the extension for (doing less would be a BC break).
- I need to figure out how to package the extension per-PHP version.
- We'd need updates here to pull the extension package for the given PHP version when the extension is specified for a workflow.

**Comment by weierophinney** [comment]
Do we want to still do this, considering the patch already merged in #107 ?

**Comment by Ocramius** [comment]
Nope, closing here 👍

**Pull request closed without merging**